app/gui2/src/util/callTree.ts

```diff
@@ -210,78 +208,133 @@ export class ArgumentApplication {
-            appTree,
-            argument: ArgumentPlaceholder.WithRetrievedConfig(
-              callId,
-              argIndex,
-              argInfo,
-              ApplicationKind.Prefix,
-              !canInsertPositional,
-              widgetCfg,
-            ),
-          })
-
-          canInsertPositional = false
-        }
-      }
+    function nextArgumentName() {
+      return tryGetIndex(knownArguments, argumentsLeftToMatch[0])?.name
+    }
+
+    function takeNextArgument() {
+      const index = argumentsLeftToMatch.shift()
+      const info = tryGetIndex(knownArguments, index)
+      return index != null && info != null ? { index, info } : undefined
+    }
+
+    function takeNamedArgument(name: string) {
+      const takeIdx = argumentsLeftToMatch.findIndex(
+        (id) => tryGetIndex(knownArguments, id)?.name === name,
+      )
+      const index = argumentsLeftToMatch.splice(takeIdx, 1)[0]
+      const info = tryGetIndex(knownArguments, index)
+      return index != null && info != null ? { index, info } : undefined
     }
 
-    for (const realArg of interpreted.args) {
-      if (realArg.argName == null) {
-        const argIndex = argumentsLeftToMatch.shift()
-        if (argIndex != null) insertPlaceholdersUpto(argIndex, realArg.appTree.function)
-        prefixArgsToDisplay.push({
-          appTree: realArg.appTree,
+    function putBackArgument(index: number) {
+      argumentsLeftToMatch.unshift(index)
+    }
+
+    const lastPositionalArgIndex = findLastIndex(interpreted.args, (arg) => arg.argName == null)
+
+    let placeholderAlreadyInserted = false
+
+    let nextArgument: ReturnType<typeof takeNextArgument>
+
+    // Always insert a placeholder for the missing argument at the first position that is legal
+    // and don't invalidate further named arguments, treating the named arguments at correct
+    // position as if they were positional. foo a b c [d]
+    for (let position = 0; position < interpreted.args.length; ++position) {
+      const arg = interpreted.args[position]
+      assert(!!arg)
+      const pastPositionalArguments = mapOr(lastPositionalArgIndex, true, (i) => position > i)
+
+      if (pastPositionalArguments && arg.argName != null && arg.argName !== nextArgumentName()) {
+        // Named argument that is not in its natural position, and there are no more
+        // positional arguments to emit in the chain. At this point placeholders can be
+        // inserted. We need to figure out which placeholders can be inserted before
+        // emitting this named argument.
+
+        // all remaining arguments must be named, as we are past all positional arguments.
+        const remainingArguments = interpreted.args.slice(position)
+
+        // For each subsequent argument in its current natural position, insert a
+        // placeholder. Do that only if the argument is not defined further in the chain.
+        while ((nextArgument = takeNextArgument())) {
```

Comment:
What would help me in this code is the clear distinction when by `argument` we mean `official argument from definition` and when `actually written argument`. Now we're both iterating over arguments and taking them from a kind of queue.

This comment confuses me, because "argument in its current natural position" suggests that those are the latter (actually written - arguments from definition are always in their "natural positions"), but the fact we insert placeholders for them implies former (args from definition - written arguments don't need a placeholder).